services/src/main/java/org/keycloak/authentication/authenticators/x509/AbstractX509ClientCertificateAuthenticator.java

```diff
@@ -129,6 +136,18 @@ protected Response createInfoResponse(AuthenticationFlowContext context, String
             }
             return null;
         };
+        
+        private static final Function<X509Certificate[], String> getSerialnumberFunc(X509AuthenticatorConfigModel config) {
+            return config.isSerialnumberHex() ? 
+                    certs -> Hex.toHexString(certs[0].getSerialNumber().toByteArray()) : 
```

Comment:
Hex.toHexString on ByteArray returns correct hex representation with leading zeroes if sign bit is 1. E.g. 00a1
toString(16) does not include leading zeros in that case.